File: Focus/Services/RedditAPIService.swift
import Foundation
import SwiftUI

// MARK: - Reddit API Service
class RedditAPIService: ObservableObject {
    static let shared = RedditAPIService()
    
    private let baseURL = "https://www.reddit.com"
    private let session = URLSession.shared
    
    @Published var posts: [RedditPost] = []
    @Published var isLoading = false
    @Published var error: String?
    
    private init() {}
    
    func searchPosts(query: String, subreddit: String? = nil, sort: RedditSort = .relevance, timeFilter: RedditTimeFilter = .all) async {
        await MainActor.run {
            isLoading = true
            error = nil
        }
        
        do {
            let searchResults = try await performSearch(query: query, subreddit: subreddit, sort: sort, timeFilter: timeFilter)
            
            await MainActor.run {
                self.posts = searchResults
                self.isLoading = false
            }
        } catch {
            await MainActor.run {
                self.error = error.localizedDescription
                self.isLoading = false
            }
        }
    }
    
    private func performSearch(query: String, subreddit: String?, sort: RedditSort, timeFilter: RedditTimeFilter) async throws -> [RedditPost] {
        var urlString = "\(baseURL)/search.json"
        var queryItems = [
            URLQueryItem(name: "q", value: query),
            URLQueryItem(name: "sort", value: sort.rawValue),
            URLQueryItem(name: "t", value: timeFilter.rawValue),
            URLQueryItem(name: "limit", value: "25"),
            URLQueryItem(name: "raw_json", value: "1")
        ]
        
        if let subreddit = subreddit {
            queryItems.append(URLQueryItem(name: "restrict_sr", value: "1"))
            urlString = "\(baseURL)/r/\(subreddit)/search.json"
        }
        
        var urlComponents = URLComponents(string: urlString)!
        urlComponents.queryItems = queryItems
        
        guard let url = urlComponents.url else {
            throw RedditAPIError.invalidURL
        }
        
        let (data, response) = try await session.data(from: url)
        
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw RedditAPIError.networkError
        }
        
        let redditResponse = try JSONDecoder().decode(RedditResponse.self, from: data)
        return redditResponse.data.children.map { $0.data }
    }
}

// MARK: - Reddit Data Models
struct RedditResponse: Codable {
    let data: RedditListing
}

struct RedditListing: Codable {
    let children: [RedditChild]
}

struct RedditChild: Codable {
    let data: RedditPost
}

struct RedditPost: Codable, Identifiable {
    let id: String
    let title: String
    let author: String
    let subreddit: String
    let score: Int
    let numComments: Int
    let created: Double
    let url: String
    let permalink: String
    let selftext: String?
    let thumbnail: String?
    let preview: RedditPreview?
    let isVideo: Bool
    
    enum CodingKeys: String, CodingKey {
        case id, title, author, subreddit, score, url, permalink, selftext, thumbnail, preview
        case numComments = "num_comments"
        case created = "created_utc"
        case isVideo = "is_video"
    }
    
    var redditURL: String {
        return "https://www.reddit.com\(permalink)"
    }
    
    var createdDate: Date {
        return Date(timeIntervalSince1970: created)
    }
    
    var thumbnailURL: URL? {
        guard let thumbnail = thumbnail,
              thumbnail != "self",
              thumbnail != "default",
              thumbnail != "nsfw",
              thumbnail != "spoiler",
              let url = URL(string: thumbnail) else {
            return nil
        }
        return url
    }
    
    var previewImageURL: URL? {
        guard let preview = preview,
              let source = preview.images.first?.source,
              let url = URL(string: source.url.replacingOccurrences(of: "&amp;", with: "&")) else {
            return nil
        }
        return url
    }
}

struct RedditPreview: Codable {
    let images: [RedditImage]
}

struct RedditImage: Codable {
    let source: RedditImageSource
}

struct RedditImageSource: Codable {
    let url: String
    let width: Int
    let height: Int
}

// MARK: - Reddit Enums
enum RedditSort: String, CaseIterable {
    case relevance = "relevance"
    case hot = "hot"
    case top = "top"
    case new = "new"
    case comments = "comments"
    
    var displayName: String {
        switch self {
        case .relevance: return "Relevance"
        case .hot: return "Hot"
        case .top: return "Top"
        case .new: return "New"
        case .comments: return "Comments"
        }
    }
}

enum RedditTimeFilter: String, CaseIterable {
    case all = "all"
    case year = "year"
    case month = "month"
    case week = "week"
    case day = "day"
    case hour = "hour"
    
    var displayName: String {
        switch self {
        case .all: return "All Time"
        case .year: return "Past Year"
        case .month: return "Past Month"
        case .week: return "Past Week"
        case .day: return "Past Day"
        case .hour: return "Past Hour"
        }
    }
}

enum RedditAPIError: Error, LocalizedError {
    case invalidURL
    case networkError
    case decodingError
    
    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid URL"
        case .networkError:
            return "Network error occurred"
        case .decodingError:
            return "Failed to decode response"
        }
    }
}

File: SkipFeedAndroid/app/src/main/java/com/skipfeed/android/data/RedditApiService.kt
package com.skipfeed.android.data

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder

@Serializable
data class RedditResponse(
    val data: RedditListing
)

@Serializable
data class RedditListing(
    val children: List<RedditChild>,
    val after: String? = null
)

@Serializable
data class RedditChild(
    val data: RedditPost
)

@Serializable
data class RedditPost(
    val id: String,
    val title: String,
    val author: String,
    val subreddit: String,
    val score: Int,
    val num_comments: Int,
    val created_utc: Double,
    val url: String,
    val permalink: String,
    val selftext: String? = null,
    val thumbnail: String? = null,
    val preview: RedditPreview? = null,
    val is_video: Boolean = false
) {
    val redditUrl: String
        get() = "https://www.reddit.com$permalink"
    
    val previewImageUrl: String?
        get() = preview?.images?.firstOrNull()?.source?.url?.replace("&amp;", "&")
    
    val thumbnailUrl: String?
        get() = if (thumbnail != null && thumbnail != "self" && thumbnail != "default") thumbnail else null
}

@Serializable
data class RedditPreview(
    val images: List<RedditImage>? = null
)

@Serializable
data class RedditImage(
    val source: RedditImageSource
)

@Serializable
data class RedditImageSource(
    val url: String
)

enum class RedditSort(val value: String, val displayName: String) {
    RELEVANCE("relevance", "Relevance"),
    HOT("hot", "Hot"),
    TOP("top", "Top"),
    NEW("new", "New"),
    COMMENTS("comments", "Comments")
}

enum class RedditTimeFilter(val value: String, val displayName: String) {
    ALL("all", "All Time"),
    YEAR("year", "Past Year"),
    MONTH("month", "Past Month"),
    WEEK("week", "Past Week"),
    DAY("day", "Past Day"),
    HOUR("hour", "Past Hour")
}

class RedditApiService {
    private val json = Json { 
        ignoreUnknownKeys = true
        coerceInputValues = true
    }
    
    suspend fun searchPosts(
        query: String,
        sort: RedditSort = RedditSort.RELEVANCE,
        timeFilter: RedditTimeFilter = RedditTimeFilter.ALL,
        after: String? = null,
        limit: Int = 25
    ): Result<Pair<List<RedditPost>, String?>> = withContext(Dispatchers.IO) {
        try {
            val encodedQuery = URLEncoder.encode(query, "UTF-8")
            val url = buildString {
                append("https://www.reddit.com/search.json")
                append("?q=$encodedQuery")
                append("&sort=${sort.value}")
                append("&t=${timeFilter.value}")
                append("&limit=$limit")
                append("&raw_json=1")
                if (after != null) {
                    append("&after=$after")
                }
            }
            
            val connection = URL(url).openConnection() as HttpURLConnection
            connection.requestMethod = "GET"
            connection.setRequestProperty("User-Agent", "SkipFeed Android App")
            
            if (connection.responseCode == 200) {
                val response = connection.inputStream.bufferedReader().use { it.readText() }
                val redditResponse = json.decodeFromString<RedditResponse>(response)
                val posts = redditResponse.data.children.map { it.data }
                val afterToken = redditResponse.data.after
                
                Result.success(Pair(posts, afterToken))
            } else {
                Result.failure(Exception("HTTP ${connection.responseCode}"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}
